platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java

```diff
@@ -0,0 +1,210 @@
+    this.restClientFactoryFactory = clientFactoryFactory;
+    properties =
+        AccessController.doPrivileged(
+            (PrivilegedAction<java.util.Properties>) System::getProperties);
+  }
+
+  private void initialize() {
+    SecureCxfClientFactory<SolrAuthResource> factory =
+        restClientFactoryFactory.getSecureCxfClientFactory(
+            properties.getProperty("solr.http.url"),
+            SolrAuthResource.class,
+            properties.getProperty("solr.username"),
+            getPlaintextPasswordFromProperties());
+    solrAuthResource = factory.getClient();
+  }
+
+  /**
+   * If the password is not changed, change it. Only call this method if the system is configured to
+   * use basic authentication for Solr. This method will only change the password if it detects that
+   * the default password is still in use. This object is intended to be a singleton, a blueprint
+   * bean. That is why is can use itself as the lock
+   */
+  public synchronized void updateSolrPassword() {
+    if (configuredToAttemptAutoPasswordChange() && isUsingDefaultPassword()) {
+      initialize();
+      generatePassword();
+      setPasswordInSolr();
+      if (isSolrPasswordChangeSuccessfull()) {
+        setPasswordInMemory();
+        setPasswordInPropertiesFile();
+      }
+    }
+  }
+
+  private void setPasswordInMemory() {
+    properties.setProperty(SOLR_PASSWORD_PROPERTY_NAME, newPasswordWrappedEncrypted);
+    LOGGER.info("Updated encrypted Solr password in memory.");
+  }
+
+  private boolean configuredToAttemptAutoPasswordChange() {
+    return Boolean.valueOf(properties.getProperty("solr.attemptAutoPasswordChange"));
+  }
+
+  /**
+   * The return value of this method will change after setPasswordInMemory() is called
+   *
+   * @return true if in-memory password is the same as hard coded default password
+   */
+  private boolean isUsingDefaultPassword() {
+    return getPlaintextPasswordFromProperties().equals("autogenerated");
+  }
+
+  private @Nullable String getPlaintextPasswordFromProperties() {
+    String property = properties.getProperty(SOLR_PASSWORD_PROPERTY_NAME);
+    if (StringUtils.isBlank(property)) {
+      LOGGER.info(
+          "Solr password system property is missing or blank. System might not be able to communicate with Solr.");
+    }
+    return encryptionService.decryptValue(property);
+  }
+
+  /**
+   * Persist the encrypted password to the system properties file. This change does not take effect
+   * until the application is restarted.
+   */
+  private void setPasswordInPropertiesFile() {
+
+    String etcDir = properties.getProperty(KARAF_ETC);
+    String systemPropertyFilename = etcDir + File.separator + SYSTEM_PROPERTIES_FILE;
+    File systemPropertiesFile = new File(systemPropertyFilename);
+    AccessController.doPrivileged(
+        (PrivilegedAction<Void>)
+            () -> {
+              try {
+                Properties systemDotProperties = new Properties(systemPropertiesFile);
```

Comment:
❓ How do we ensure that only one service at a time is modifying this file? If more than one service is doing that for different (or the same) reasons, then this file would end up being corrupted or with missing entries.

Reply:
🤔

Reply:
This is done as part of starting up whenever a client is created. There is no guarantee of when that will happen.